### PR TITLE
[gala_kinematic] Masking the loss for the first observation

### PR DESCRIPTION
**Pull request by vincentpierre**
## Motivation and Context
Masking the losses using the mask in the rollout.
This mask corresponds to the first observation after a reset which needs to be discarded

## Types of changes

- Docs change / refactoring / dependency upgrade
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
Hmm, all the tests including the easily fixable ones like Black are failing on these tests though. 👎

**Comment by vincentpierre** [comment]
>Hmm, all the tests including the easily fixable ones like Black are failing on these tests though. 👎 

They are failing on the target branch. This change is not responsible

File: habitat_baselines/rl/ppo/ppo.py [hunk 1]
```diff
@@ -142,6 +142,16 @@ def update(self, rollouts: RolloutStorage) -> Tuple[float, float, float]:
                     else:
                         value_loss = 0.5 * (batch["returns"] - values).pow(2)
 
+
+                    # Mask the loss. The mask corresponds to the transition right
+                    # after a reset. This first observation needs to be ignored
+                    # because its observation is actually part of the previous episode
+                    action_loss = action_loss * batch["masks"]
```
[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_baselines/rl/ppo/ppo.py:149]: `torch.where(masks, action_loss, action_loss.new_zeros(())` was marginally faster last time I profiled this but that certainly isn't the bottleneck so no big deal.

More importantly, @vincentpierre I don't think we can re-use masks here. Masks signifies the start of the episode but that isn't quite what this is, you are masking out the episode reset step. And actually masks will need to be true twice per episode for gala. Once to stop the bootstrap of the returns and once to reset the hidden state before the true first step. For a rollout, we'd need something like

```
Time step :                  T - 1  |  T | <reset> | 0 | 1
Not Done Masks:        True  | True   | False | False | True
Is Reset Step:              False | False | True  | False  | False
```

[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_baselines/rl/ppo/ppo.py:149]: More importantly, @vincentpierre I don't think we can re-use masks here. Masks signifies the start of the episode but that isn't quite what this is, you are masking out the episode reset step. And actually masks will need to be true twice per episode for gala. Once to stop the bootstrap of the returns and once to reset the hidden state before the true first step. For a rollout, we'd need something like


| Time step/Field | T - 1  |  T |  RESET  | 0 | 1 |
| --------------- | ------| -- | -------- | - | - | 
| Not Done Masks | True  | True   | False | False | True |
| Is Reset Step | False | False | True  | False  | False |
